nodego/supervisor.go

```diff
@@ -0,0 +1,315 @@
+		if err := logBatch.report(); err != nil {
+			fmt.Fprintln(os.Stderr, err.Error())
+			killInstance()
+		}
+	}
+}
+
+func (ctx *loggingContext) initialize() {
+	ctx.initOnce.Do(func() {
+		ctx.queue = make(chan *logBatch, 5)
+		ctx.startNewBatch()
+		go ctx.startReportWorker()
+	})
+}
+
+var loggingCtx loggingContext
+
+var (
+	infoLogWriter  supervisorWriter = "INFO"
+	errorLogWriter supervisorWriter = "ERROR"
+
+	// InfoLogger is a logger that batches sends logs to the supervisor with a
+	// severity level of INFO.
+	InfoLogger = log.New(infoLogWriter, "", 0)
+	// InfoLogger is a logger that batches sends logs to the supervisor with a
+	// severity level of ERROR.
+	ErrorLogger = log.New(errorLogWriter, "", 0)
+)
+
+func init() {
+	if supervisorHostname != "" && supervisorInternalPort != "" {
+		loggingCtx.initialize()
+	}
+}
+
+const isoTimeFormat = "2006-01-02T15:04:05.999Z07:00"
+
+type supervisorWriter string
+
+// Write implements io.Writer.Write.
+func (w supervisorWriter) Write(p []byte) (int, error) {
+	entry := &logEntry{
+		TextPayload: string(p),
+		Severity:    string(w),
+		Time:        time.Now().Format(isoTimeFormat),
+		ExecutionID: loggingCtx.executionID(),
+	}
+
+	if !loggingCtx.addEntry(entry) {
+		return os.Stderr.Write(entry.consoleOutput())
+	}
+
+	return len(entry.TextPayload), nil
+}
+
+func newSupervisorRequest(path string, v interface{}) (*http.Request, error) {
+	postData, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", (&url.URL{
+		Scheme: "http",
+		Host:   supervisorHostname + ":" + supervisorInternalPort,
+		Path:   path,
+	}).String(), bytes.NewBuffer(postData))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Content-Length", strconv.Itoa(len(postData)))
+
+	return req, nil
+}
+
+func doRequestWithContext(ctx context.Context, r *http.Request) (*http.Response, error) {
+	resp, err := http.DefaultClient.Do(r.WithContext(ctx))
+	if err != nil {
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+		default:
+		}
+	}
+	return resp, err
+}
+
+func postToSupervisor(path string, v interface{}, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	req, err := newSupervisorRequest(path, v)
+	if err != nil {
+		return err
+	}
+
+	resp, err := doRequestWithContext(ctx, req)
+	if err == ctx.Err() {
+		return errors.New("timeout when calling supervisor")
+	} else if err != nil {
+		return fmt.Errorf("error when calling supervisor: %s\n\n%s\n", err.Error(), debug.Stack())
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		err = fmt.Errorf("incorrect response code from supervisor: %d\n", resp.StatusCode)
+	}
+
+	return err
+}
+
+func killInstance() {
+	err := postToSupervisor("/_ah/kill", nil, supervisorKillTimeout)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+	}
+	// Exit code 16 is copied over from worker.js
```

Comment:
One last style thing: even these comments ending in worker.js should end in a period.